pandas/core/internals/concat.py

```diff
@@ -471,7 +471,11 @@ def get_reindexed_values(self, empty_dtype: DtypeObj, upcasted_na) -> ArrayLike:
                     if len(values) and values[0] is None:
                         fill_value = None
 
-                if is_datetime64tz_dtype(empty_dtype):
+                if is_dtype_equal(blk_dtype, empty_dtype) and self.indexers:
```

Comment:
is there a reason this is caught here, instead of in the `elif is_1d_only_ea_dtype` block. In the OP you mention the benefits for EA types. What are the benefits for using this path for the builtin dtypes? (and for a targeted backport?)

Reply:
I can move it into the `is_1d_only_ea_dtype` block. That would fix it for the specific use case of custom extension dtypes. 

Looking a bit more in detail, if we have a numpy dtype, the approach with np.empty might actually faster than `take`:

```
In [6]: arr = np.array([np.nan]*10000)

In [7]: indexer = np.arange(10000, dtype="int32")

In [9]: %timeit pd.core.algorithms.take_nd(arr, indexer, axis=0)
28.4 µs ± 752 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)

In [11]: %%timeit
    ...: arr2 = np.empty(arr.shape, dtype=arr.dtype)
    ...: arr2.fill(np.nan)
5.68 µs ± 135 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
```

So that's a good reason to keep the check specific for extension dtypes.